### PR TITLE
CoreDNS default DNS provider for k8s 1.14 and up

### PR DESCRIPTION
**Pull request by superseb**
https://github.com/rancher/rancher/issues/19522

### CODE REVIEW

File: cluster/defaults.go [hunk 1]
```diff
@@ -273,6 +276,32 @@ func (c *Cluster) setClusterImageDefaults() error {
 	return nil
 }
 
+func (c *Cluster) setClusterDNSDefaults() error {
+	if c.DNS == nil || len(c.DNS.Provider) == 0 {
```
Comment (moelsayed) [inline cluster/defaults.go]: any chance we can reverse this check and return nil as early as possible ?
``` 
if c.DNS != nil && len(c.DNS.Provider) != 0 {
   return nil
}
```

File: cluster/defaults.go [hunk 2]
```diff
@@ -273,6 +276,32 @@ func (c *Cluster) setClusterImageDefaults() error {
...
+		}
+		logrus.Debugf("No DNS provider configured, setting default based on cluster version [%s]", clusterSemVer)
+		K8sVersionCoreDNSSemVer, err := util.StrToSemVer(K8sVersionCoreDNS)
+		if err != nil {
+			return err
+		}
+		// Default DNS provider for cluster version 1.14.0 and higher is coredns
+		ClusterDNSProvider := CoreDNSProvider
+		// If cluster version is less than 1.14.0 (K8sVersionCoreDNSSemVer), use kube-dns
+		if clusterSemVer.LessThan(*K8sVersionCoreDNSSemVer) {
```
Comment (moelsayed) [inline cluster/defaults.go]: also might be better reversed, use the Default and override it at the condition.

Comment (superseb) [reply cluster/defaults.go]: Due to a vendoring issue, we have an older `go-semver` lib with any `LessThan` available. I can reverse it by using `! if clusterSemVer.LessThan` but that seemed more ugly than this. If you still rather have that I can change it.

The vendoring issue will take more time to investigate and fix appropriately (and backport)

Comment (moelsayed) [reply cluster/defaults.go]: Not worth it. It was just a nit.

**Pull request merged**